### ISSUE TITLE
Potențial infinite recursion on updateModalState

### ISSUE THREAD

**Issue by seanmdalton** [jokob-sk/NetAlertX#923]
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues and I checked the docs https://github.com/jokob-sk/NetAlertX/tree/main/docs

### The issue occurs in the following browsers. Select at least 2.

- [ ] Firefox
- [X] Chrome
- [ ] Edge
- [ ] Safari (unsupported) - PRs welcome
- [ ] N/A - This is an issue with the backend

### Current Behavior

When only loading the main Devices page, I noticed the Loading modal spinning for an excessive period of time (30+ seconds). When I pulled up the browser console (Google Chrome), I saw 8k (and quickly counting) requests to `/php/server/query_logs.php?file=execution_queue.log&nocache=1735577383548` before my browser crashed.  

When inspecting these calls, I see the following output:
```
[2024-12-30_11-37-05]|26551019-7584-4d2f-acc5-dd59341504c9|update_api|devices,appevents
[2024-12-30_11-37-29]|615f2daf-e97d-4ab6-8928-394d14431c1f|update_api|devices,appevents
```

Upon digging in (and pardon some naivety here), I'm seeing an infinitely recursive function that seems to be the cause: updateModalStat() 

I threw a 'return true;' before the setTimeOut function/ajax call, and it stopped entirely. Page loads fine now (but of course lacking whatever this code block is doing). 

```
// --------------------------------------------------------              
// Updating the execution queue in in modal pop-up                              
function updateModalState() {                           
  return true;                  //  <----- I ADDED THIS FOR TESTING
  setTimeout(function() {                                                                  
      // Fetch the content from the log file using an AJAX request              
      $.ajax({                                                       
          url: '/php/server/query_logs.php?file=execution_queue.log',
          type: 'GET',                                                          
          success: function(data) {                                                       
              // Update the content of the HTML element (e.g., a div with id 'logContent')
              $('#'+modalEventStatusId).html(data);                             
                                                      
              updateModalState();                                                                                    
          },                             
          error: function() {                                            
              // Handle error, such as the file not being found                                                                                                             
              $('#logContent').html('Error: Log file not found.');
          }                                       
      });                           
  }, 2000);                           
}         
```

I don't fully grok the intention of this code chain/block, so I don't want to presume a fix, but I can provide any other data you think would be helpful. 

I'm running Netalert in Docker, on a macvlan network, behind Traefik/TLS, and I haven't noticed any other problems. Every other page loads perfectly fine/fast. 

### Expected Behavior

Since I don't know the intentions of the code block, I'm not sure what the expected behavior to be. That said, I'd expect there to be potentially a pause between calls to limit load on the service, some break condition in the recursive stack, and this to not block (by not removing the modal) the load of the rest of the page. 

### Steps To Reproduce

1. load /devices.php 
2. open dev console
3. see lots of calls? 

### What branch are you running?

Production

### Debug enabled

- [X] I have read and followed the steps in the wiki link above and provided the required debug logs and the log section covers the time when the issue occurs.

**Comment by seanmdalton** [comment]
Just kidding. This didn't actually fix anything. I'll keep digging

**Comment by seanmdalton** [comment]
Maybe I understand a bit more of what's happening. Potentially an execution queue contention? I can easily replicate the issue by changing a node attribute (just by updating the name, for example), clicking save, and immediately jumping back to the Devices page. This might be coinciding with a scan execution. 

looking a bit more I see the actual calls are coming from 'app/front/devices.php', 
```
function handleLoadingDialog(needsReload = false)
{                     
                           
  // console.log('needsReload:');  
  // console.log(needsReload);
                                         
  $.get('/php/server/query_logs.php?file=execution_queue.log&nocache=' + Date.now(), function(data) {

    if(data.includes("update_api|devices"))                                     
    {
      showSpinner("devices_old")
                                                                                
      setTimeout(handleLoadingDialog(true), 1000);

    } else if (needsReload)                                                     
    {                               
      location.reload(); 
    }else                             
    {                                
      hideSpinner();
    }         

  })                                                                            
                             
}
```

and when I watch the execution log, I'll see the 'update_api|devices' line in the file. Once I see that line drop from the log, the page loads. 

There doesn't seem to be anything useful in the debug logs but I saved some examples if it'll help.

**Comment by jokob-sk** [comment]
Hey @seanmdalton ,

Thanks a lot spending this much time digging into this. I'm not 100% sure whats going on, but I'll explain how this is supposed to work. Happy to improve if you have ideas or even accept a PR :) 

1. When a device is edited the api endpoint has to be updated
2. The edit triggers an update api event by adding the `[2024-12-30_11-37-29]|615f2daf-e97d-4ab6-8928-394d14431c1f|update_api|devices,appevents` line into the `execution_queue.log`
3. The backend checks the `execution_queue.log` every 5s (or after a delay is a scan is in progress as the scan may update the API as well). 
4. After the API endpoint is updated the `execution_queue.log` is cleared. 
5. The modal on frontend updates the state of the displayed message 


Maybe this is too convoluted, so happy to hear about possible improvements.

**Comment by netExtra** [comment]
> No activity for 2 weeks - closing for now

I just had the update API loop. It lasted about 2-3 minutes and would occasionally cycle through other states like AVHISCAN but always back to Update API. It was triggered by a multi-edit of 2 "new devices". The Alert Events, Alert Down boxes were clear/empty but I clicked the save option anyway to make sure that these devices did not report any events. The "new device" box was ticked so I unticked that, and then change Network Node.

**Comment by jokob-sk** [comment]
Hi @netExtra ,

Thanks for getting back. 

Did the Loading dialog eventually disappear?

It might have been expected behavior with backend processes locking the DB and teh UI waiting fro API commits to finish.

**Comment by netExtra** [comment]
> Hi [@netExtra](https://github.com/netExtra) ,
> 
> Thanks for getting back.
> 
> Did the Loading dialog eventually disappear?
> 
> It might have been expected behavior with backend processes locking the DB and teh UI waiting fro API commits to finish.

Yes. As I said it lasted about 2-3 minutes. Expected behaviour? OK. But 2-3 minutes is way too much time.

**Comment by jokob-sk** [comment]
Hi @netExtra ,

tried something. Can you please switch to the `netalertx-dev` docker image (backup everything at first or, ideally spin up a **separate container with a separate db and config**), in about 15 minutes (or after the last [action finishes](https://github.com/jokob-sk/NetAlertX/actions)) from now and test it?

thanks,
j

**Comment by netExtra** [comment]
> Hi [@netExtra](https://github.com/netExtra) ,
> 
> tried something. Can you please switch to the `netalertx-dev` docker image (backup everything at first or, ideally spin up a **separate container with a separate db and config**), in about 15 minutes (or after the last [action finishes](https://github.com/jokob-sk/NetAlertX/actions)) from now and test it?
> 
> thanks, j

Hi. I was already using dev :) I will update to the latest image.

**Comment by jokob-sk** [comment]
@netExtra thanks for checking. 

I'll keep this open until the next release in case this reoccurs. In short, the app is now also triggering api updates after each scan, not only after all scans finish, so even if there's a wait time, it should be shorter. Let's see.

**Comment by netExtra** [comment]
> [@netExtra](https://github.com/netExtra) thanks for checking.
> 
> I'll keep this open until the next release in case this reoccurs. In short, the app is now also triggering api updates after each scan, not only after all scans finish, so even if there's a wait time, it should be shorter. Let's see.

Thanks. I think Netalertx is the main bottleneck on my small resource starved server. I'm thinking that maybe my Internet is running better because Linux server is running better because Netalertx is running better. So if you made changes to improve efficiency then maybe it's working :)

**Comment by jokob-sk** [comment]
BTW I just setup [Beszel](https://github.com/henrygd/beszel) for server monitoring to see if my changes increase NAX resource usage - you might find it useful for monitoring as well.

**Comment by ryanwinter** [comment]
I'm also seeing this issue, thanks for the explanation on what is going on behind the scenes because this was driving me crazy.  So now I know that it can take up to 5 seconds when I update device details before the UI will respond again, and also that a scan will block the update as well.

Also turns out the the AVAHISCAN is running alot and blocking things, so I'm going to investigate what is going on there :)

**Comment by ryanwinter** [comment]
Turned on logging and it turns out the AVAHISCAN is running every minute, and I have 5 unknown devices it scans, each of which times out after 10 seconds. So yeah this is pretty much continuously running.

The `AVAHISCAN_RUN` is set to `before_name_updates`, but I can't find information on what that value actually means.

**Comment by jokob-sk** [comment]
@ryanwinter I've updated the save message on a device to highlight that there might be a delay. Let me know if this is clear enough:

![Image](https://github.com/user-attachments/assets/4cf90f1a-58fe-40c2-b8d7-64d63e8324ad)

**Comment by jokob-sk** [comment]
You can find info on the "before_name_updates" setting here: https://github.com/jokob-sk/NetAlertX/blob/9d8b147e40ca09a41b8f8636cd9463d6a2a94d23/docs/PLUGINS_DEV.md#supported-settings-function-values

**Comment by ryanwinter** [comment]
Thanks @jokob-sk, that screenshot looks great! 

For the `before_name_updates`, I did find that page but was still confused about what it meant. I ended up looking through the source code and main loop that runs every minute, and the description describes the position in the main loop :)

**Comment by jokob-sk** [comment]
Feel free to suggest a better description and I'll update the docs
accordingly.
